### PR TITLE
Make for-loops work in the Limited C-API

### PR DESCRIPTION
**Pull request by scoder**
Closes https://github.com/cython/cython/issues/5641

### CODE REVIEW

File: tests/run/isolated_limited_api_tests.srctree [hunk 1]
```diff
@@ -48,6 +53,13 @@ def fib(int n):
         a, b = b, a + b
     return b
 
+@cython.binding(False)
```
Comment (da-woods) [inline tests/run/isolated_limited_api_tests.srctree:56]: We shouldn't need `binding(False)` (although it doesn't hurt to have it either)

**Review by da-woods — COMMENTED** [review]
Looks good. One comment about the tests but that can easily be ignored too.

**Comment by da-woods** [comment]
> Looks good.

I say that without spotting that the tests don't pass...

**Comment by scoder** [comment]
Surprisingly, the tests only crash when run in parallel, not when run sequentially. I'll try to investigate that.

I'm also not entirely happy with the generated code because the no-safe-macros case could be implemented in a much simpler way without distinguishing lists and tuples, just using the sequence API (once we know that we have a list or tuple). But the way the current C code is generated makes that difficult and prone to redundancy…

I wonder if the "optimise for lists and tuples" case is really relevant for the Limited API case. Maybe we can just use the fallback code that grabs the iterator and be done…

**Comment by da-woods** [comment]
> I wonder if the "optimise for lists and tuples" case is really relevant for the Limited API case. Maybe we can just use the fallback code that grabs the iterator and be done…

My wild guess is that the people who are interested in limited API are mostly those aiming to wrap some C library rather than those interested in performance. And thus they'd probably prioritise "working" ahead of optimized. But I have no evidence for that.

**Pull request merged**